iotilecore/iotile/core/hw/debug/firmware_image_analyzer.py

```diff
@@ -0,0 +1,150 @@
+            self.get_cdb_app_info()
+
+        return (self.cdb_app_info['api_major_version'],
+                self.cdb_app_info['api_minor_version'])
+
+    def get_firmware_name(self):
+        """Returns firmware name specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return self.cdb_app_info['name'].decode("utf-8")
+
+    def get_firmware_version(self):
+        """Returns firmware module version specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return (self.cdb_app_info['module_major_version'],
+                self.cdb_app_info['module_minor_version'],
+                self.cdb_app_info['module_patch_version'])
+
+    def get_cdb_app_info(self):
+        """Gets the string of bytes of the CDB app info"""
+        cdb_app_info_bytes = self._hex_image.gets(self.max_addr - (FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO - 1), \
+                FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+        if self._check_cdb_app_info(cdb_app_info_bytes):
+            self.cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+            return self.cdb_app_info
+
+        addresses = self._hex_image.addresses()
+        for address_index in range(0, self.num_addresses - 4):
+            search_bytes = self._hex_image.gets(addresses[address_index], 4)
+            search, = struct.unpack("<L", search_bytes)
+            if search == self.magic_number:
+                cdb_app_info_bytes = self._hex_image.gets(addresses[address_index - FirmwareImageAnalyzer.MAGIC_NUMBER_BYTE_OFFSET], \
+                    FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+                if self._check_cdb_app_info(cdb_app_info_bytes):
+                    self.cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+                    return self.cdb_app_info
+                
+
+    def _check_cdb_app_info(self, cdb_app_info_bytes):
+        """Helper function to check the byte string if it is a valid CDB app info"""
+        _hardware_type, _api_major_version, _api_minor_version, name, _module_major_version, \
+            _module_minor_version, _module_patch_version, _num_slave_commands, _num_required_configs, \
+            _num_total_configs, _reserved, _p_config_variables, _p_slave_handlers, magic_number, \
+            _firmware_checksum = struct.unpack("<BBB6sBBBBBBBLLLL", cdb_app_info_bytes)
+        cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+
+        try:
+            name = cdb_app_info['name'].decode("utf-8")
+        except UnicodeDecodeError:
+            return False
+        except:
+            raise ExternalError("Unexpected error resolving the firmware name.")
+
+        if name in FirmwareImageAnalyzer.KNOWN_FIRMWARE_NAMES \
+            and cdb_app_info['magic_number'] == self.magic_number:
+            return True
+
+        return False
+
+    def _format_cdb_app_info(self, cdb_app_info_bytes):
```

Comment:
Style points.  This can just be a standalone function with the same name at the bottom of the module.  Standalone functions can be better because they're easier to test (you don't need to create the class, just call the function) and from an understanding point of view, they make the class smaller so it's easier to get a sense of what it does.

Reply:
Fixed in next push.